### PR TITLE
Add UNet3+ model

### PR DESCRIPTION
**Pull request by geoyee**
参考github-pytorch代码（https://github.com/ZJUGiveLab/UNet-Version），经过测试（修改bisenet_optic_disc_512x512_1k.yml中的model的type），可以train的起来

### CODE REVIEW

File: paddleseg/models/unet_3plus.py [hunk 1]
```diff
@@ -0,0 +1,716 @@
...
+import paddle.nn as nn
+import paddle.nn.functional as F
+
+from paddleseg.cvlibs import manager
+from paddleseg.models.layers.layer_libs import unetConv2, SyncBatchNorm
+from paddleseg.cvlibs.param_init import kaiming_normal_init
+
+
+@manager.MODELS.add_component
+class UNet_3Plus(nn.Layer):
```
Comment (michaelowenliu) [inline paddleseg/models/unet_3plus.py]: 类命名 首字母大写，无其他符号

File: paddleseg/models/unet_3plus.py [hunk 2]
```diff
@@ -0,0 +1,716 @@
...
+    """
+    The UNet3+ implementation based on PaddlePaddle.
+
+    The original article refers to
+    Huang H , Lin L , Tong R , et al. UNet 3+: A Full-Scale Connected UNet for Medical Image Segmentation[J]. arXiv, 2020.
+    (https://arxiv.org/ftp/arxiv/papers/2004/2004.08790.pdf).
+
+    Args:
+        in_channels (int): The channel number of input image.  Default: 3.
```
Comment (michaelowenliu) [inline paddleseg/models/unet_3plus.py]: 注释规范，如果有默认值，则应该写（int, optional）

File: paddleseg/models/unet_3plus.py [hunk 3]
```diff
@@ -0,0 +1,716 @@
...
+    """
+    def __init__(self, in_channels=3, num_classes=2, feature_scale=4, is_deconv=True, is_batchnorm=True, end_sigmoid=False):
+        super(UNet_3Plus, self).__init__()
+        self.is_deconv = is_deconv
+        self.in_channels = in_channels
+        self.is_batchnorm = is_batchnorm
+        self.feature_scale = feature_scale
+        self.end_sigmoid = end_sigmoid
+        filters = [64, 128, 256, 512, 1024]
+        ## -------------Encoder--------------
```
Comment (michaelowenliu) [inline paddleseg/models/unet_3plus.py]: 组网方式的可读性不高，拆解成组件化，Encoder和Decoder

File: paddleseg/models/layers/layer_libs.py [hunk 4]
```diff
@@ -165,3 +165,37 @@ def forward(self, x):
         x = self.dropout(x)
         x = self.conv(x)
         return x
+
+
+class unetConv2(nn.Layer):
```
Comment (michaelowenliu) [inline paddleseg/models/layers/layer_libs.py]: 不在内置库里面添加新方式，放在自定义模型文件里，class命名方式参考下面review意见。

File: paddleseg/models/unet_3plus.py [hunk 5]
```diff
@@ -0,0 +1,716 @@
...
+        d1 = self.outconv1(hd1)  # d1->320*320*num_classes
+        if self.end_sigmoid:
+            out = [F.sigmoid(d1)]
+        else:
+            out = [d1]
+        return out
+
+
+@manager.MODELS.add_component
+class UNet_3Plus_DeepSup(nn.Layer):
```
Comment (michaelowenliu) [inline paddleseg/models/unet_3plus.py]: 同样存在命名规范问题

File: paddleseg/models/unet_3plus.py [hunk 6]
```diff
@@ -0,0 +1,716 @@
...
+    """
+    The UNet3+ with deep supervision implementation based on PaddlePaddle.
+
+    The original article refers to
+    Huang H , Lin L , Tong R , et al. UNet 3+: A Full-Scale Connected UNet for Medical Image Segmentation[J]. arXiv, 2020.
+    (https://arxiv.org/ftp/arxiv/papers/2004/2004.08790.pdf).
+
+    Args:
+        in_channels (int): The channel number of input image.  Default: 3.
```
Comment (michaelowenliu) [inline paddleseg/models/unet_3plus.py]: 注释规范问题

File: paddleseg/models/unet_3plus.py [hunk 7]
```diff
@@ -0,0 +1,716 @@
...
+    """
+    def __init__(self, in_channels=3, num_classes=2, feature_scale=4, is_deconv=True, is_batchnorm=True, end_sigmoid=False):
+        super(UNet_3Plus_DeepSup, self).__init__()
+        self.is_deconv = is_deconv
+        self.in_channels = in_channels
+        self.is_batchnorm = is_batchnorm
+        self.feature_scale = feature_scale
+        self.end_sigmoid = end_sigmoid
+        filters = [64, 128, 256, 512, 1024]
+        ## -------------Encoder--------------
```
Comment (michaelowenliu) [inline paddleseg/models/unet_3plus.py]: 组网可读性问题

**Review by michaelowenliu — CHANGES_REQUESTED** [review]
提了一些review意见

**Comment by geoyee** [comment]
收到，非常感谢回复和指导，这是我第一次在开源项目中提交PR，下面我将在我的空闲时间中根据建议继续完善！


On 3/1/2021 12:05，Liu Yi<notifications@github.com> wrote：

@michaelowenliu requested changes on this pull request.

提了一些review意见

In paddleseg/models/unet_3plus.py:

> +# distributed under the License is distributed on an "AS IS" BASIS,

+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.

+# See the License for the specific language governing permissions and

+# limitations under the License.

+

+import paddle

+import paddle.nn as nn

+import paddle.nn.functional as F

+

+from paddleseg.cvlibs import manager

+from paddleseg.models.layers.layer_libs import unetConv2, SyncBatchNorm

+from paddleseg.cvlibs.param_init import kaiming_normal_init

+

+

+@manager.MODELS.add_component

+class UNet_3Plus(nn.Layer):



类命名 首字母大写，无其他符号

In paddleseg/models/unet_3plus.py:

> +from paddleseg.cvlibs import manager

+from paddleseg.models.layers.layer_libs import unetConv2, SyncBatchNorm

+from paddleseg.cvlibs.param_init import kaiming_normal_init

+

+

+@manager.MODELS.add_component

+class UNet_3Plus(nn.Layer):

+    """

+    The UNet3+ implementation based on PaddlePaddle.

+

+    The original article refers to

+    Huang H , Lin L , Tong R , et al. UNet 3+: A Full-Scale Connected UNet for Medical Image Segmentation[J]. arXiv, 2020.

+    (https://arxiv.org/ftp/arxiv/papers/2004/2004.08790.pdf).

+

+    Args:

+        in_channels (int): The channel number of input image.  Default: 3.



注释规范，如果有默认值，则应该写（int, optional）

In paddleseg/models/unet_3plus.py:

> +        num_classes (int): The unique number of target classes.  Default: 2.

+        feature_scale (int): The feature map scale.  Default: 4.

+        is_deconv (bool): A bool value indicates whether using deconvolution in upsampling.

+            If False, use resize_bilinear. Default: True.

+        is_batchnorm (bool): use batchnorm after conv or not.  Default: True.

+        end_sigmoid (bool): use sigmoid for results or not. Default: False.

+    """

+    def __init__(self, in_channels=3, num_classes=2, feature_scale=4, is_deconv=True, is_batchnorm=True, end_sigmoid=False):

+        super(UNet_3Plus, self).__init__()

+        self.is_deconv = is_deconv

+        self.in_channels = in_channels

+        self.is_batchnorm = is_batchnorm

+        self.feature_scale = feature_scale

+        self.end_sigmoid = end_sigmoid

+        filters = [64, 128, 256, 512, 1024]

+        ## -------------Encoder--------------



组网方式的可读性不高，拆解成组件化，Encoder和Decoder

In paddleseg/models/layers/layer_libs.py:

> @@ -165,3 +165,37 @@ def forward(self, x):

         x = self.dropout(x)

         x = self.conv(x)

         return x

+

+

+class unetConv2(nn.Layer):



不在内置库里面添加新方式，放在自定义模型文件里，class命名方式参考下面review意见。

In paddleseg/models/unet_3plus.py:

> +        hd2_UT_hd1 = self.hd2_UT_hd1_relu(self.hd2_UT_hd1_bn(self.hd2_UT_hd1_conv(self.hd2_UT_hd1(hd2))))

+        hd3_UT_hd1 = self.hd3_UT_hd1_relu(self.hd3_UT_hd1_bn(self.hd3_UT_hd1_conv(self.hd3_UT_hd1(hd3))))

+        hd4_UT_hd1 = self.hd4_UT_hd1_relu(self.hd4_UT_hd1_bn(self.hd4_UT_hd1_conv(self.hd4_UT_hd1(hd4))))

+        hd5_UT_hd1 = self.hd5_UT_hd1_relu(self.hd5_UT_hd1_bn(self.hd5_UT_hd1_conv(self.hd5_UT_hd1(hd5))))

+        hd1 = self.relu1d_1(self.bn1d_1(self.conv1d_1(

+            paddle.concat([h1_Cat_hd1, hd2_UT_hd1, hd3_UT_hd1, hd4_UT_hd1, hd5_UT_hd1], 1)))) # hd1->320*320*UpChannels

+        d1 = self.outconv1(hd1)  # d1->320*320*num_classes

+        if self.end_sigmoid:

+            out = [F.sigmoid(d1)]

+        else:

+            out = [d1]

+        return out

+

+

+@manager.MODELS.add_component

+class UNet_3Plus_DeepSup(nn.Layer):



同样存在命名规范问题

In paddleseg/models/unet_3plus.py:

> +        else:

+            out = [d1]

+        return out

+

+

+@manager.MODELS.add_component

+class UNet_3Plus_DeepSup(nn.Layer):

+    """

+    The UNet3+ with deep supervision implementation based on PaddlePaddle.

+

+    The original article refers to

+    Huang H , Lin L , Tong R , et al. UNet 3+: A Full-Scale Connected UNet for Medical Image Segmentation[J]. arXiv, 2020.

+    (https://arxiv.org/ftp/arxiv/papers/2004/2004.08790.pdf).

+

+    Args:

+        in_channels (int): The channel number of input image.  Default: 3.



注释规范问题

In paddleseg/models/unet_3plus.py:

> +        num_classes (int): The unique number of target classes.  Default: 2.

+        feature_scale (int): The feature map scale.  Default: 4.

+        is_deconv (bool): A bool value indicates whether using deconvolution in upsampling.

+            If False, use resize_bilinear. Default: True.

+        is_batchnorm (bool): use batchnorm after conv or not.  Default: True.

+        end_sigmoid (bool): use sigmoid for results or not. Default: False.

+    """

+    def __init__(self, in_channels=3, num_classes=2, feature_scale=4, is_deconv=True, is_batchnorm=True, end_sigmoid=False):

+        super(UNet_3Plus_DeepSup, self).__init__()

+        self.is_deconv = is_deconv

+        self.in_channels = in_channels

+        self.is_batchnorm = is_batchnorm

+        self.feature_scale = feature_scale

+        self.end_sigmoid = end_sigmoid

+        filters = [64, 128, 256, 512, 1024]

+        ## -------------Encoder--------------



组网可读性问题

**Comment by haoyuying** [comment]
> 收到，非常感谢回复和指导，这是我第一次在开源项目中提交PR，下面我将在我的空闲时间中根据建议继续完善！ On 3/1/2021 12:05，Liu Yi<notifications@github.com> wrote： @michaelowenliu requested changes on this pull request. 提了一些review意见 In paddleseg/models/unet_3plus.py:
> +# distributed under the License is distributed on an "AS IS" BASIS,
> +# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. +# See the License for the specific language governing permissions and +# limitations under the License. + +import paddle +import paddle.nn as nn +import paddle.nn.functional as F + +from paddleseg.cvlibs import manager +from paddleseg.models.layers.layer_libs import unetConv2, SyncBatchNorm +from paddleseg.cvlibs.param_init import kaiming_normal_init + + +@manager.MODELS.add_component +class UNet_3Plus(nn.Layer): 类命名 首字母大写，无其他符号 In paddleseg/models/unet_3plus.py:
> +from paddleseg.cvlibs import manager
> +from paddleseg.models.layers.layer_libs import unetConv2, SyncBatchNorm +from paddleseg.cvlibs.param_init import kaiming_normal_init + + +@manager.MODELS.add_component +class UNet_3Plus(nn.Layer): + """ + The UNet3+ implementation based on PaddlePaddle. + + The original article refers to + Huang H , Lin L , Tong R , et al. UNet 3+: A Full-Scale Connected UNet for Medical Image Segmentation[J]. arXiv, 2020. + (https://arxiv.org/ftp/arxiv/papers/2004/2004.08790.pdf). + + Args: + in_channels (int): The channel number of input image. Default: 3. 注释规范，如果有默认值，则应该写（int, optional） In paddleseg/models/unet_3plus.py:
> + num_classes (int): The unique number of target classes. Default: 2.
> + feature_scale (int): The feature map scale. Default: 4. + is_deconv (bool): A bool value indicates whether using deconvolution in upsampling. + If False, use resize_bilinear. Default: True. + is_batchnorm (bool): use batchnorm after conv or not. Default: True. + end_sigmoid (bool): use sigmoid for results or not. Default: False. + """ + def __init__(self, in_channels=3, num_classes=2, feature_scale=4, is_deconv=True, is_batchnorm=True, end_sigmoid=False): + super(UNet_3Plus, self).__init__() + self.is_deconv = is_deconv + self.in_channels = in_channels + self.is_batchnorm = is_batchnorm + self.feature_scale = feature_scale + self.end_sigmoid = end_sigmoid + filters = [64, 128, 256, 512, 1024] + ## -------------Encoder-------------- 组网方式的可读性不高，拆解成组件化，Encoder和Decoder In paddleseg/models/layers/layer_libs.py:
> @@ -165,3 +165,37 @@ def forward(self, x):
> x = self.dropout(x) x = self.conv(x) return x + + +class unetConv2(nn.Layer): 不在内置库里面添加新方式，放在自定义模型文件里，class命名方式参考下面review意见。 In paddleseg/models/unet_3plus.py:
> + hd2_UT_hd1 = self.hd2_UT_hd1_relu(self.hd2_UT_hd1_bn(self.hd2_UT_hd1_conv(self.hd2_UT_hd1(hd2))))
> + hd3_UT_hd1 = self.hd3_UT_hd1_relu(self.hd3_UT_hd1_bn(self.hd3_UT_hd1_conv(self.hd3_UT_hd1(hd3)))) + hd4_UT_hd1 = self.hd4_UT_hd1_relu(self.hd4_UT_hd1_bn(self.hd4_UT_hd1_conv(self.hd4_UT_hd1(hd4)))) + hd5_UT_hd1 = self.hd5_UT_hd1_relu(self.hd5_UT_hd1_bn(self.hd5_UT_hd1_conv(self.hd5_UT_hd1(hd5)))) + hd1 = self.relu1d_1(self.bn1d_1(self.conv1d_1( + paddle.concat([h1_Cat_hd1, hd2_UT_hd1, hd3_UT_hd1, hd4_UT_hd1, hd5_UT_hd1], 1)))) # hd1->320*320*UpChannels + d1 = self.outconv1(hd1) # d1->320*320*num_classes + if self.end_sigmoid: + out = [F.sigmoid(d1)] + else: + out = [d1] + return out + + +@manager.MODELS.add_component +class UNet_3Plus_DeepSup(nn.Layer): 同样存在命名规范问题 In paddleseg/models/unet_3plus.py:
> + else:
> + out = [d1] + return out + + +@manager.MODELS.add_component +class UNet_3Plus_DeepSup(nn.Layer): + """ + The UNet3+ with deep supervision implementation based on PaddlePaddle. + + The original article refers to + Huang H , Lin L , Tong R , et al. UNet 3+: A Full-Scale Connected UNet for Medical Image Segmentation[J]. arXiv, 2020. + (https://arxiv.org/ftp/arxiv/papers/2004/2004.08790.pdf). + + Args: + in_channels (int): The channel number of input image. Default: 3. 注释规范问题 In paddleseg/models/unet_3plus.py:
> + num_classes (int): The unique number of target classes. Default: 2.
> + feature_scale (int): The feature map scale. Default: 4. + is_deconv (bool): A bool value indicates whether using deconvolution in upsampling. + If False, use resize_bilinear. Default: True. + is_batchnorm (bool): use batchnorm after conv or not. Default: True. + end_sigmoid (bool): use sigmoid for results or not. Default: False. + """ + def __init__(self, in_channels=3, num_classes=2, feature_scale=4, is_deconv=True, is_batchnorm=True, end_sigmoid=False): + super(UNet_3Plus_DeepSup, self).__init__() + self.is_deconv = is_deconv + self.in_channels = in_channels + self.is_batchnorm = is_batchnorm + self.feature_scale = feature_scale + self.end_sigmoid = end_sigmoid + filters = [64, 128, 256, 512, 1024] + ## -------------Encoder-------------- 组网可读性问题 — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or unsubscribe.

请问这个模型自己在我们提供的数据集上训练过吗，可以达到多少精度呢？

**Comment by geoyee** [comment]
没有，我尝试一下使用快速入门中的案例进行测试


On 3/1/2021 20:10，haoyuying<notifications@github.com> wrote：

收到，非常感谢回复和指导，这是我第一次在开源项目中提交PR，下面我将在我的空闲时间中根据建议继续完善！ On 3/1/2021 12:05，Liu Yinotifications@github.com wrote： @michaelowenliu requested changes on this pull request. 提了一些review意见 In paddleseg/models/unet_3plus.py:
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. +# See the License for the specific language governing permissions and +# limitations under the License. + +import paddle +import paddle.nn as nn +import paddle.nn.functional as F + +from paddleseg.cvlibs import manager +from paddleseg.models.layers.layer_libs import unetConv2, SyncBatchNorm +from paddleseg.cvlibs.param_init import kaiming_normal_init + + +@manager.MODELS.add_component +class UNet_3Plus(nn.Layer): 类命名 首字母大写，无其他符号 In paddleseg/models/unet_3plus.py:
+from paddleseg.cvlibs import manager
+from paddleseg.models.layers.layer_libs import unetConv2, SyncBatchNorm +from paddleseg.cvlibs.param_init import kaiming_normal_init + + +@manager.MODELS.add_component +class UNet_3Plus(nn.Layer): + """ + The UNet3+ implementation based on PaddlePaddle. + + The original article refers to + Huang H , Lin L , Tong R , et al. UNet 3+: A Full-Scale Connected UNet for Medical Image Segmentation[J]. arXiv, 2020. + (https://arxiv.org/ftp/arxiv/papers/2004/2004.08790.pdf). + + Args: + in_channels (int): The channel number of input image. Default: 3. 注释规范，如果有默认值，则应该写（int, optional） In paddleseg/models/unet_3plus.py:

num_classes (int): The unique number of target classes. Default: 2.
feature_scale (int): The feature map scale. Default: 4. + is_deconv (bool): A bool value indicates whether using deconvolution in upsampling. + If False, use resize_bilinear. Default: True. + is_batchnorm (bool): use batchnorm after conv or not. Default: True. + end_sigmoid (bool): use sigmoid for results or not. Default: False. + """ + def init(self, in_channels=3, num_classes=2, feature_scale=4, is_deconv=True, is_batchnorm=True, end_sigmoid=False): + super(UNet_3Plus, self).init() + self.is_deconv = is_deconv + self.in_channels = in_channels + self.is_batchnorm = is_batchnorm + self.feature_scale = feature_scale + self.end_sigmoid = end_sigmoid + filters = [64, 128, 256, 512, 1024] + ## -------------Encoder-------------- 组网方式的可读性不高，拆解成组件化，Encoder和Decoder In paddleseg/models/layers/layer_libs.py:
@@ -165,3 +165,37 @@ def forward(self, x):
x = self.dropout(x) x = self.conv(x) return x + + +class unetConv2(nn.Layer): 不在内置库里面添加新方式，放在自定义模型文件里，class命名方式参考下面review意见。 In paddleseg/models/unet_3plus.py:
hd2_UT_hd1 = self.hd2_UT_hd1_relu(self.hd2_UT_hd1_bn(self.hd2_UT_hd1_conv(self.hd2_UT_hd1(hd2))))
hd3_UT_hd1 = self.hd3_UT_hd1_relu(self.hd3_UT_hd1_bn(self.hd3_UT_hd1_conv(self.hd3_UT_hd1(hd3)))) + hd4_UT_hd1 = self.hd4_UT_hd1_relu(self.hd4_UT_hd1_bn(self.hd4_UT_hd1_conv(self.hd4_UT_hd1(hd4)))) + hd5_UT_hd1 = self.hd5_UT_hd1_relu(self.hd5_UT_hd1_bn(self.hd5_UT_hd1_conv(self.hd5_UT_hd1(hd5)))) + hd1 = self.relu1d_1(self.bn1d_1(self.conv1d_1( + paddle.concat([h1_Cat_hd1, hd2_UT_hd1, hd3_UT_hd1, hd4_UT_hd1, hd5_UT_hd1], 1)))) # hd1->320320UpChannels + d1 = self.outconv1(hd1) # d1->320320num_classes + if self.end_sigmoid: + out = [F.sigmoid(d1)] + else: + out = [d1] + return out + + +@manager.MODELS.add_component +class UNet_3Plus_DeepSup(nn.Layer): 同样存在命名规范问题 In paddleseg/models/unet_3plus.py:
else:
out = [d1] + return out + + +@manager.MODELS.add_component +class UNet_3Plus_DeepSup(nn.Layer): + """ + The UNet3+ with deep supervision implementation based on PaddlePaddle. + + The original article refers to + Huang H , Lin L , Tong R , et al. UNet 3+: A Full-Scale Connected UNet for Medical Image Segmentation[J]. arXiv, 2020. + (https://arxiv.org/ftp/arxiv/papers/2004/2004.08790.pdf). + + Args: + in_channels (int): The channel number of input image. Default: 3. 注释规范问题 In paddleseg/models/unet_3plus.py:
num_classes (int): The unique number of target classes. Default: 2.
feature_scale (int): The feature map scale. Default: 4. + is_deconv (bool): A bool value indicates whether using deconvolution in upsampling. + If False, use resize_bilinear. Default: True. + is_batchnorm (bool): use batchnorm after conv or not. Default: True. + end_sigmoid (bool): use sigmoid for results or not. Default: False. + """ + def init(self, in_channels=3, num_classes=2, feature_scale=4, is_deconv=True, is_batchnorm=True, end_sigmoid=False): + super(UNet_3Plus_DeepSup, self).init() + self.is_deconv = is_deconv + self.in_channels = in_channels + self.is_batchnorm = is_batchnorm + self.feature_scale = feature_scale + self.end_sigmoid = end_sigmoid + filters = [64, 128, 256, 512, 1024] + ## -------------Encoder-------------- 组网可读性问题 — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or unsubscribe.

请问这个模型自己在我们提供的数据集上训练过吗，可以达到多少精度呢？

**Comment by haoyuying** [comment]
感谢您对paddlehub的贡献，如果您能够提供一个较优的预训练模型结果，将对我们的repo更具有价值和意义。

**Comment by geoyee** [comment]
参考paddleseg的入门指南（https://gitee.com/paddlepaddle/PaddleSeg/blob/release/v2.0.0-rc/docs/quick_start.md），只在配置中改了一下模型，其他参数保持不变训练的过程及结果在此项目中（https://aistudio.baidu.com/aistudio/projectdetail/1591568）


On 3/1/2021 21:04，haoyuying<notifications@github.com> wrote：

感谢您对paddlehub的贡献，如果您能够提供一个较优的预训练模型结果，将对我们的repo更具有价值和意义。

File: paddleseg/models/unet_3plus.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (haoyuying) [inline paddleseg/models/unet_3plus.py:1]: 改成2021吧

File: paddleseg/models/unet_3plus.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+        # initialise weights
+        for sublayer in self.sublayers ():
+            if isinstance(sublayer, nn.Conv2D):
+                kaiming_normal_init(sublayer.weight)
+            elif isinstance(sublayer, (nn.BatchNorm, nn.SyncBatchNorm)):
+                kaiming_normal_init(sublayer.weight)
+    def dotProduct(self, seg, cls):
+        B, N, H, W = seg.shape
+        seg = seg.reshape((B, N, H * W))
+        # final = torch.einsum("ijk,ij->ijk", [seg, cls])  # replace it with the following three lines
```
Comment (haoyuying) [inline paddleseg/models/unet_3plus.py:72]: 把这个删掉吧

File: paddleseg/models/unet_3plus.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+        self.conv1 = UnetConv2D(in_channels, filters[0], is_batchnorm)
+        self.maxpool1 = nn.MaxPool2D(kernel_size=2)
+        self.conv2 = UnetConv2D(filters[0], filters[1], is_batchnorm)
+        self.maxpool2 = nn.MaxPool2D(kernel_size=2)
+        self.conv3 = UnetConv2D(filters[1], filters[2], is_batchnorm)
+        self.maxpool3 = nn.MaxPool2D(kernel_size=2)
+        self.conv4 = UnetConv2D(filters[2], filters[3], is_batchnorm)
+        self.maxpool4 = nn.MaxPool2D(kernel_size=2)
+        self.conv5 = UnetConv2D(filters[3], filters[4], is_batchnorm)
+    def forward(self, inputs):
```
Comment (haoyuying) [inline paddleseg/models/unet_3plus.py:106]: forward之前空一行

File: paddleseg/models/unet_3plus.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+        h4 = self.conv4(h4)  # h4->40*40*512
+        h5 = self.maxpool4(h4)
+        hd5 = self.conv5(h5)  # h5->20*20*1024
+        return [h1, h2, h3, h4, hd5]
+
+
+class Decoder(nn.Layer):
+    def __init__(self, filters, cat_channels, up_channels):
+        super(Decoder, self).__init__()
+        '''stage 4d'''
```
Comment (haoyuying) [inline paddleseg/models/unet_3plus.py:122]: 每个stage 空一行吧，不然显得代码太长

File: paddleseg/models/unet_3plus.py [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+        # hd5->20*20, hd1->320*320, Upsample 16 times
+        self.hd5_UT_hd1 = nn.Upsample(scale_factor=16, mode='bilinear')  # 14*14
+        self.hd5_UT_hd1_conv = nn.Conv2D(filters[4], cat_channels, 3, padding=1)
+        self.hd5_UT_hd1_bn = nn.BatchNorm(cat_channels)
+        self.hd5_UT_hd1_relu = nn.ReLU()
+        # fusion(h1_Cat_hd1, hd2_UT_hd1, hd3_UT_hd1, hd4_UT_hd1, hd5_UT_hd1)
+        self.conv1d_1 = nn.Conv2D(up_channels, up_channels, 3, padding=1)  # 16
+        self.bn1d_1 = nn.BatchNorm(up_channels)
+        self.relu1d_1 = nn.ReLU()
+    def forward(self, inputs):
```
Comment (haoyuying) [inline paddleseg/models/unet_3plus.py:238]: 类里面的函数每个要空一行

File: paddleseg/models/layers/layer_libs.py [hunk 13]
```diff
@@ -12,16 +12,14 @@
 # See the License for the specific language governing permissions and
```
Comment (haoyuying) [inline paddleseg/models/layers/layer_libs.py:12]: 最好拉一下paddleseg最新的代码，再将您的代码添加到新拉的代码中，最好不要对seg内置的代码进行更改，以防对seg的整体功能产生影响。

**Pull request closed without merging**